### ISSUE TITLE
find_equivalent_units returns wrong result for number density

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#16124]
### Description

When asking for equivalent units for a number density, I am getting units of length.

### Expected behavior

I would have thought that any unit in the returned output should be number density units.

### How to Reproduce

```python
In [23]: u.Unit('m**-3').find_equivalent_units()
Out[23]: 
  Primary name | Unit definition | Aliases                         
[
  AU           | 1.49598e+11 m   | au, astronomical_unit            ,
  Angstrom     | 1e-10 m         | AA, angstrom                     ,
  cm           | 0.01 m          | centimeter                       ,
  earthRad     | 6.3781e+06 m    | R_earth, Rearth                  ,
  jupiterRad   | 7.1492e+07 m    | R_jup, Rjup, R_jupiter, Rjupiter ,
  lsec         | 2.99792e+08 m   | lightsecond                      ,
  lyr          | 9.46073e+15 m   | lightyear                        ,
  m            | irreducible     | meter                            ,
  micron       | 1e-06 m         |                                  ,
  pc           | 3.08568e+16 m   | parsec                           ,
  solRad       | 6.957e+08 m     | R_sun, Rsun                      ,
]
```

### Versions

```
macOS-14.1.1-arm64-arm-64bit
Python 3.11.2 (main, Mar 23 2023, 23:28:33) [Clang 14.0.0 (clang-1400.0.29.202)]
astropy 6.0.0
Numpy 1.26.0
pyerfa 2.0.0.3
Scipy 1.10.1
Matplotlib 3.6.3
```

**Comment by pllim** [comment]
After discussions with @adrn @bmorris3 @nstarman et al. on Slack, is this still considered a bug?

**Comment by mhvk** [comment]
Seems definitely a bug to me, since it is not the same for other powers:
```
u.Unit('m**3').find_equivalent_units()
Out[5]: 
  Primary name | Unit definition | Aliases 
[
  l            | 0.001 m3        | L, liter ,
]
```

Fix in #16127

**Comment by mhvk** [comment]
p.s. I think it is correct to get nothing: we have no explicit unit in our namespace that corresponds to a number density.
